src/rulesets/Base/indexing.jl

```diff
@@ -26,3 +26,43 @@ function rrule(::typeof(getindex), x::Array{<:Number}, inds...)
 
     return y, getindex_pullback
 end
+
+
+#####
+##### `eachslice` and friends
+#####
+
+function rrule(::typeof(eachrow), x::AbstractVecOrMat)
+    allrows(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, 1))
+    return collect(eachrow(x)), allrows
```

Comment:
What's the reason for changing the primal and using `collect` here (and below, of course)?

Reply:
I just copied Zygote here, but I see that last time around, I said this:

> Without it, the gradient is (sometimes) some weird NamedTuple containing fields of the generator, and I got tired of trying to sort that out.

Reply:
Yeah, I assumed it is more annoying. But on the other hand changing the primal to return something of a different type seems quite surprising and can break code that relies on the output being a generator (or, possibly in the future a special array type: https://github.com/JuliaLang/julia/pull/32310). Hence I assumed changing the computation of the primal is OK but changing its type should be avoided.

Reply:
Yes when that lands we can remove this. That PR is an equally breaking change, if you care about something being a Generator, but IMO this is super-rare: it returns something you can iterate, and code which does so ought to be fine.

Reply:
Sure, although I think AD quite often collects them for you, gives you an array gradient for a generator. Everything allocates a ton. If you call eachcol on a matrix, the rule allocates N^2 to avoid allocating N^3 without it... plus N for collecting the generator.